### PR TITLE
[TOSA] Fix aten.view and aten.slice.tensor

### PR DESCRIPTION
**Pull request by AmosLewis**
Find this error in https://github.com/nod-ai/SHARK/issues/494
Deal with aten.view input -1
Deal with aten.slice start -1

slice op explaination: https://cran.r-project.org/web/packages/torch/vignettes/indexing.html

### CODE REVIEW

**Comment by AmosLewis** [comment]
```
func.func @torch.aten.slice(%arg0: !torch.vtensor<[1,128,2],f32>) -> !torch.vtensor<[1,1,2],f32> {
  %int-1 = torch.constant.int -1
  %int1 = torch.constant.int 1
  %int0 = torch.constant.int 0
  %0 = torch.aten.slice.Tensor %arg0, %int1, %int-1, %int0, %int1 : !torch.vtensor<[1,128,2],f32>, !torch.int, !torch.int, !torch.int, !torch.int -> !torch.vtensor<[1,1,2],f32>
  return %0 : !torch.vtensor<[1,1,2],f32>
}
```

`➜  ~ torch-mlir-opt -convert-torch-to-tosa /tmp/slice.mlir `
```
module {
  func.func @torch.aten.slice(%arg0: !torch.vtensor<[1,128,2],f32>) -> !torch.vtensor<[1,1,2],f32> {
    %0 = torch_c.to_builtin_tensor %arg0 : !torch.vtensor<[1,128,2],f32> -> tensor<1x128x2xf32>
    %int-1 = torch.constant.int -1
    %int1 = torch.constant.int 1
    %int0 = torch.constant.int 0
    %1 = "tosa.slice"(%0) {size = array<i64: 1, 1, 2>, start = array<i64: 0, 127, 0>} : (tensor<1x128x2xf32>) -> tensor<1x1x2xf32>
    %2 = torch_c.from_builtin_tensor %1 : tensor<1x1x2xf32> -> !torch.vtensor<[1,1,2],f32>
    return %2 : !torch.vtensor<[1,1,2],f32>
  }
}
```

**Comment by AmosLewis** [comment]
```
func.func @torch.aten.view(%arg0: !torch.vtensor<[1,128],si64>) -> !torch.vtensor<[1,128],si64> {
  %int-1 = torch.constant.int -1
  %int128 = torch.constant.int 128
  %0 = torch.prim.ListConstruct %int-1, %int128 : (!torch.int, !torch.int) -> !torch.list<int>
  %1 = torch.aten.view %arg0, %0 : !torch.vtensor<[1,128],si64>, !torch.list<int> -> !torch.vtensor<[1,128],si64>
  return %1 : !torch.vtensor<[1,128],si64>
}
```
`➜  ~ torch-mlir-opt -convert-torch-to-tosa /tmp/view_torchbackend.mlir      `                                           
```
module {
  func.func @torch.aten.view(%arg0: !torch.vtensor<[1,128],si64>) -> !torch.vtensor<[1,128],si64> {
    %0 = torch_c.to_builtin_tensor %arg0 : !torch.vtensor<[1,128],si64> -> tensor<1x128xi64>
    %int-1 = torch.constant.int -1
    %int128 = torch.constant.int 128
    %1 = torch.prim.ListConstruct %int-1, %int128 : (!torch.int, !torch.int) -> !torch.list<int>
    %2 = "tosa.reshape"(%0) {new_shape = array<i64: 1, 128>} : (tensor<1x128xi64>) -> tensor<1x128xi64>
    %3 = torch_c.from_builtin_tensor %2 : tensor<1x128xi64> -> !torch.vtensor<[1,128],si64>
    return %3 : !torch.vtensor<[1,128],si64>
  }
}
```

**Comment by AmosLewis** [comment]
```
➜  ~ torch-mlir-opt -pass-pipeline='builtin.module(torch-backend-to-tosa-backend-pipeline)'   /tmp/view_torchbackend.mlir
module {
  func.func @torch.aten.view(%arg0: tensor<1x128xi64>) -> tensor<1x128xi64> {
    return %arg0 : tensor<1x128xi64>
  }
}

➜  ~ torch-mlir-opt -pass-pipeline='builtin.module(torch-backend-to-tosa-backend-pipeline)'   /tmp/slice.mlir            
module {
  func.func @torch.aten.slice(%arg0: tensor<1x128x2xf32>) -> tensor<1x1x2xf32> {
    %0 = "tosa.slice"(%arg0) {size = array<i64: 1, 1, 2>, start = array<i64: 0, 127, 0>} : (tensor<1x128x2xf32>) -> tensor<1x1x2xf32>
    return %0 : tensor<1x1x2xf32>
  }
}

```

File: lib/Conversion/TorchToTosa/TorchToTosa.cpp [hunk 1]
```diff
@@ -2596,6 +2596,24 @@ LogicalResult ConvertAtenOp<AtenViewOp>::matchAndRewrite(
     return rewriter.notifyMatchFailure(op,
                                        "size must consist of Scalar constants");
 
+  // # the size -1 is inferred from other dimensions
```
Comment (ramiro050) [inline lib/Conversion/TorchToTosa/TorchToTosa.cpp:2599]: Can you add e2e tests that test the new additions?

**Review by ramiro050 — CHANGES_REQUESTED** [review]

File: lib/Conversion/TorchToTosa/TorchToTosa.cpp (hunk 1 above)
Comment (AmosLewis) [reply lib/Conversion/TorchToTosa/TorchToTosa.cpp:2599]: I just add a view test in basic.mlir. Would it be enough?

Comment (ramiro050) [reply lib/Conversion/TorchToTosa/TorchToTosa.cpp:2599]: That's not enough. See: https://github.com/llvm/torch-mlir/blob/main/docs/architecture.md#dos-and-donts-for-unit-vs-end-to-end-testing

> DO use an [end-to-end test](https://github.com/llvm/torch-mlir/blob/main/docs/adding_an_e2e_test.md) if you are implementing a new op or extending the support for an existing op.

Also, the tests in basic.mlir are not needed.

Comment (AmosLewis) [reply lib/Conversion/TorchToTosa/TorchToTosa.cpp:2648]: Ok, just delete the file in basic.mlir. And add an e2e test `ViewNegativeStaticModule_basic`

Comment (AmosLewis) [reply lib/Conversion/TorchToTosa/TorchToTosa.cpp:2648]: Done

File: lib/Conversion/TorchToTosa/TorchToTosa.cpp [hunk 2]
```diff
@@ -3061,13 +3079,18 @@ LogicalResult ConvertAtenOp<AtenSliceTensorOp>::matchAndRewrite(
   if (!matchPattern(op.getStart(), m_TorchConstantInt(&start)))
     return rewriter.notifyMatchFailure(op, "start must be a Scalar constant");
 
-  if (start < 0)
```
Comment (ramiro050) [inline lib/Conversion/TorchToTosa/TorchToTosa.cpp:3064]: Can you add an e2e test that checks these changes?

File: test/Conversion/TorchToTosa/basic.mlir [hunk 3]
```diff
@@ -1065,6 +1065,25 @@ func.func @torch.aten.masked_fill.Tensor(%arg0: !torch.vtensor<[1,12,128,128],f3
...
+// CHECK-SAME:                                %[[VAL_0:.*]]: !torch.vtensor<[1,128,2],f32>) -> !torch.vtensor<[1,1,2],f32> {
+// CHECK:           %[[VAL_1:.*]] = torch_c.to_builtin_tensor %[[VAL_0]] : !torch.vtensor<[1,128,2],f32> -> tensor<1x128x2xf32>
+// CHECK:           %[[VAL_2:.*]] = torch.constant.int -1
+// CHECK:           %[[VAL_3:.*]] = torch.constant.int 1
+// CHECK:           %[[VAL_4:.*]] = torch.constant.int 0
+// CHECK:           %[[VAL_5:.*]] = "tosa.slice"(%[[VAL_1]]) {size = array<i64: 1, 1, 2>, start = array<i64: 0, 127, 0>} : (tensor<1x128x2xf32>) -> tensor<1x1x2xf32>
+// CHECK:           %[[VAL_6:.*]] = torch_c.from_builtin_tensor %[[VAL_5]] : tensor<1x1x2xf32> -> !torch.vtensor<[1,1,2],f32>
+// CHECK:           return %[[VAL_6]] : !torch.vtensor<[1,1,2],f32>
+// CHECK:         }
+func.func @torch.aten.slice(%arg0: !torch.vtensor<[1,128,2],f32>) -> !torch.vtensor<[1,1,2],f32> {
```
Comment (ramiro050) [inline test/Conversion/TorchToTosa/basic.mlir:1079]: Not needed

File: lib/Conversion/TorchToTosa/TorchToTosa.cpp [hunk 4]
```diff
@@ -2645,6 +2645,24 @@ LogicalResult ConvertAtenOp<AtenViewOp>::matchAndRewrite(
     return rewriter.notifyMatchFailure(op,
                                        "size must consist of Scalar constants");
 
+  // # the size -1 is inferred from other dimensions
```
Comment (ramiro050) [inline lib/Conversion/TorchToTosa/TorchToTosa.cpp:2648]: You need to first make sure there is at most one `-1` in the list and return a `notifyMatchFailure` if that is not the case

File: lib/Conversion/TorchToTosa/TorchToTosa.cpp [hunk 5]
```diff
@@ -2645,6 +2645,24 @@ LogicalResult ConvertAtenOp<AtenViewOp>::matchAndRewrite(
...
+  }
+  size_t otherSize = 1;
+  for (size_t i = 0; i < outShape.size(); i++) {
+    if (outShape[i] > 0) {
+      otherSize *= outShape[i];
+    }
+  }
+  for (size_t i = 0; i < outShape.size(); i++) {
+    if (outShape[i] < 0) {
+      outShape[i] = totalSize / otherSize;
```
Comment (ramiro050) [inline lib/Conversion/TorchToTosa/TorchToTosa.cpp:2662]: I would add a `break` after this line to make it very clear that this is expected to only runs once

File: lib/Conversion/TorchToTosa/TorchToTosa.cpp [hunk 6]
```diff
@@ -3061,13 +3079,18 @@ LogicalResult ConvertAtenOp<AtenSliceTensorOp>::matchAndRewrite(
   if (!matchPattern(op.getStart(), m_TorchConstantInt(&start)))
     return rewriter.notifyMatchFailure(op, "start must be a Scalar constant");
 
-  if (start < 0)
-    return rewriter.notifyMatchFailure(op, "Currently unsupported: start < 0");
+  if (start < 0) {
+    start = start + selfType.getShape()[dim];
```
Comment (ramiro050) [inline lib/Conversion/TorchToTosa/TorchToTosa.cpp:3083]: You need to check that it is positive after this

File: lib/Conversion/TorchToTosa/TorchToTosa.cpp [hunk 7]
```diff
@@ -3061,13 +3079,18 @@ LogicalResult ConvertAtenOp<AtenSliceTensorOp>::matchAndRewrite(
...
+  }
 
   int64_t end;
   if (!matchPattern(op.getEnd(), m_TorchConstantInt(&end)))
     return rewriter.notifyMatchFailure(op, "end must be a Scalar constant");
 
+  if (end <= 0) {
```
Comment (ramiro050) [inline lib/Conversion/TorchToTosa/TorchToTosa.cpp:3090]: I don't think this is correct. For the `end == 0` case, `end` should remain zero. Make sure to also e2e test this edge case

**Review by ramiro050 — CHANGES_REQUESTED** [review]

File: lib/Conversion/TorchToTosa/TorchToTosa.cpp (hunk 7 above)
Comment (AmosLewis) [reply lib/Conversion/TorchToTosa/TorchToTosa.cpp:3090]: I no 0, then in the case end=0, start=-1, it will bug. The end < start will comeout. But when end=0, start=-1, it tries to get the last value.

Comment (AmosLewis) [reply lib/Conversion/TorchToTosa/TorchToTosa.cpp:3090]: I tried to delete it but:
```
Trying to match "(anonymous namespace)::ConvertAtenOp<mlir::torch::Torch::AtenSliceTensorOp>"
    ** Failure : Currently unsupported: end < start
"(anonymous namespace)::ConvertAtenOp<mlir::torch::Torch::AtenSliceTensorOp>" result 0
  } -> FAILURE : pattern failed to match
} -> FAILURE : no matched legalization pattern
```

File: lib/Conversion/TorchToTosa/TorchToTosa.cpp (hunk 6 above)
Comment (AmosLewis) [reply lib/Conversion/TorchToTosa/TorchToTosa.cpp:3083]: Done

File: lib/Conversion/TorchToTosa/TorchToTosa.cpp (hunk 5 above)
Comment (AmosLewis) [reply lib/Conversion/TorchToTosa/TorchToTosa.cpp:2662]: Done

File: lib/Conversion/TorchToTosa/TorchToTosa.cpp (hunk 4 above)
Comment (AmosLewis) [reply lib/Conversion/TorchToTosa/TorchToTosa.cpp:2648]: Done

**Comment by AmosLewis** [comment]
Split the aten.view op into a new patch https://github.com/llvm/torch-mlir/pull/1815
